actors/builtin/miner/deadline_state_test.go

```diff
@@ -0,0 +1,379 @@
+package miner_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/builtin/miner"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/filecoin-project/specs-actors/support/mock"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeadlines(t *testing.T) {
+	sectors := []*miner.SectorOnChainInfo{
+		testSector(2, 1, 50, 60, 1000),
+		testSector(3, 2, 51, 61, 1001),
+		testSector(7, 3, 52, 62, 1002),
+		testSector(8, 4, 53, 63, 1003),
+
+		testSector(11, 5, 54, 64, 1004),
+		testSector(13, 6, 55, 65, 1005),
+		testSector(8, 7, 56, 66, 1006),
+		testSector(8, 8, 57, 67, 1007),
+
+		testSector(8, 9, 58, 68, 1008),
+	}
+
+	sectorSize := abi.SectorSize(32 << 30)
+	quantSpec := miner.NewQuantSpec(4, 1)
+	partitionSize := uint64(4)
+	builder := mock.NewBuilder(context.Background(), address.Undef)
+
+	t.Run("adds sectors", func(t *testing.T) {
+		rt := builder.Build(t)
+		store := adt.AsStore(rt)
+
+		dl := emptyDeadline(t, rt)
+		power, err := dl.AddSectors(store, partitionSize, sectors, sectorSize, quantSpec)
+		require.NoError(t, err)
+
+		expectedPower := miner.PowerForSectors(sectorSize, sectors)
+		assert.True(t, expectedPower.Equals(power))
+		assertDeadlineState(t, rt,
+			dl,
+			sectors,
+			sectorSize,
+			partitionSize,
+			bf(), // faults
+			bf(), // recovering
+			bf(), // terminations
+			bf(), // posts
+
+			bf(1, 2, 3, 4),
+			bf(5, 6, 7, 8),
+			bf(9),
+		)
+	})
+
+	t.Run("terminates sectors", func(t *testing.T) {
+		rt := builder.Build(t)
+		store := adt.AsStore(rt)
+
+		dl := emptyDeadline(t, rt)
+
+		// Add sectors.
+		{
+			_, err := dl.AddSectors(store, partitionSize, sectors, sectorSize, quantSpec)
+			require.NoError(t, err)
+
+			assertDeadlineState(t, rt,
+				dl,
+				sectors,
+				sectorSize,
+				partitionSize,
+				bf(), // faults
+				bf(), // recovering
+				bf(), // terminations
+				bf(), // posts
+
+				bf(1, 2, 3, 4),
+				bf(5, 6, 7, 8),
+				bf(9),
+			)
+		}
+
+		// Terminate sectors in partitions 0 and 1, mark sector 9 faulty.
+		// Unfortunately, we don't have per-deadline methods for stuff
+		// like this.
```

Comment:
We could improve this a bit, though, right? At worst with free functions that take and mutate both deadline and partition state. Improving things like this on the way to testing is worthwhile.

Reply:
Good point. We'd have to export them but that's probably not a huge issue.

Reply:
https://github.com/filecoin-project/specs-actors/pull/761